File: server/bcephal-grid/src/main/java/com/moriset/bcephal/grid/domain/Grille.java
/**
 * 
 */
package com.moriset.bcephal.grid.domain;

import java.util.ArrayList;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonSubTypes.Type;
import com.moriset.bcephal.domain.DataSourceType;
import com.moriset.bcephal.domain.ListChangeHandler;
import com.moriset.bcephal.domain.MainObject;
import com.moriset.bcephal.domain.dimension.DimensionType;
import com.moriset.bcephal.domain.filters.UniverseFilter;

import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Inheritance;
import jakarta.persistence.InheritanceType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.PostLoad;
import jakarta.persistence.SequenceGenerator;
import jakarta.persistence.Table;
import jakarta.persistence.Transient;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;

/**
 * @author B-Cephal Team
 *
 *         29 mars 2021
 */
@Inheritance(strategy = InheritanceType.JOINED)
@JsonSubTypes({ @Type(value = Grille.class, name = "Grille") })
@Entity(name = "Grille")
@Table(name = "BCP_GRID")
@Data
@EqualsAndHashCode(callSuper = false)
public class Grille extends MainObject {

	private static final long serialVersionUID = -4558463610223968350L;

	@Id
	@GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "grid_seq")
	@SequenceGenerator(name = "grid_seq", sequenceName = "grid_seq", initialValue = 1,  allocationSize = 1)
	private Long id;	
	
	@Enumerated(EnumType.STRING)
	private DataSourceType dataSourceType;
	
	private Long dataSourceId;
	
	@Transient
	private String dataSourceName;

	@Enumerated(EnumType.STRING) 
	private GrilleType type;
		
	@Enumerated(EnumType.STRING) 
	private GrilleCategory category;

	@Enumerated(EnumType.STRING) 
	private GrilleStatus status;
	
	@Enumerated(EnumType.STRING) 
	private GrilleSource sourceType;
	
	private Long sourceId;
	
	private Long reportId;
	
	private boolean editable;	
	
	private boolean showAllRowsByDefault;
	
	private boolean allowLineCounting;
	
	private boolean consolidated;
		
	//private boolean useLink;
	
	@ToString.Exclude @EqualsAndHashCode.Exclude
	@ManyToOne @JoinColumn(name = "userFilter")
	private UniverseFilter userFilter;	
	
	@ToString.Exclude @EqualsAndHashCode.Exclude
	@ManyToOne @JoinColumn(name = "adminFilter")
	private UniverseFilter adminFilter;
	
	@ToString.Exclude @EqualsAndHashCode.Exclude
	@Transient @JsonIgnore
	private UniverseFilter gridUserFilter;
	@ToString.Exclude @EqualsAndHashCode.Exclude
	@Transient @JsonIgnore
	private UniverseFilter gridAdminFilter;
	
	private boolean debit;

	private boolean credit;

	@Enumerated(EnumType.STRING) 
	private GrilleRowType rowType;
	
	private Boolean published;
	
	private Integer visibleColumnCount;
	
	
	@ToString.Exclude @EqualsAndHashCode.Exclude
	@JsonIgnore
	@OneToMany(cascade = jakarta.persistence.CascadeType.REFRESH, fetch = FetchType.LAZY, mappedBy = "grid")
	private List<GrilleDimension> dimensions;

	@ToString.Exclude @EqualsAndHashCode.Exclude
	@Transient
	private ListChangeHandler<GrilleDimension> dimensionListChangeHandler;
	
	@ToString.Exclude @EqualsAndHashCode.Exclude
	@JsonIgnore
	@OneToMany(cascade = jakarta.persistence.CascadeType.REFRESH, fetch = FetchType.LAZY, mappedBy = "grid")
	private List<GrilleColumn> columns;

	@ToString.Exclude @EqualsAndHashCode.Exclude
	@Transient 
	private ListChangeHandler<GrilleColumn> columnListChangeHandler;
	

	/**
	 * Default constructor
	 */
	public Grille() {
		this.columnListChangeHandler = new ListChangeHandler<GrilleColumn>();		
		this.dimensionListChangeHandler = new ListChangeHandler<GrilleDimension>();
		editable = false;
		this.category = GrilleCategory.USER;
		this.status = GrilleStatus.LOADED;
		this.sourceType = GrilleSource.USER;
		this.dataSourceType = DataSourceType.UNIVERSE;
		this.visibleColumnCount = 5;
		this.published = false;
	}
	
	public Grille(Long id) {
		this();
		this.id = id;
	}
	
	public int getVisibleColumnCount(){
		if(visibleColumnCount == null) {
			visibleColumnCount = 5;
		}
		return visibleColumnCount;
	}
	
	public boolean isUseLink() {
		return false;
	}
	
	public DataSourceType getDataSourceType(){
		if(dataSourceType == null) {
			dataSourceType = DataSourceType.UNIVERSE;
		}
		return dataSourceType;
	}
	
	public Boolean getPublished() {
		if(published == null) {
			published = false;
		}
		return published;
	}
	
	public void setDimensions(List<GrilleDimension> dimensions) {
		this.dimensions = dimensions;
		dimensionListChangeHandler.setOriginalList(dimensions);
	}

	public void setColumns(List<GrilleColumn> columns) {
		this.columns = columns;
		columnListChangeHandler.setOriginalList(columns);
	}
		
	
	@JsonIgnore
	public boolean isInput() {
		return getType() != null && getType() == GrilleType.INPUT;
	}
	
	@JsonIgnore
	public boolean isReport() {
		return getType() != null && getType() == GrilleType.REPORT;
	}
	
	@JsonIgnore
	public boolean isReconciliation() {
		return getType() != null && getType() == GrilleType.RECONCILIATION;
	}
	
	@JsonIgnore
	public boolean isBillingEventRepo() {
		return getType() != null && getType() == GrilleType.BILLING_EVENT_REPOSITORY;
	}
	
	@JsonIgnore
	public boolean isClientRepo() {
		return getType() != null && getType() == GrilleType.CLIENT_REPOSITORY;
	}
	
	@JsonIgnore
	public boolean isInvoiceRepo() {
		return getType() != null && getType() == GrilleType.INVOICE_REPOSITORY;
	}
	
	@JsonIgnore
	public boolean isCreditNoteRepo() {
		return getType() != null && getType() == GrilleType.CREDIT_NOTE_REPOSITORY;
	}
	
	@JsonIgnore
	public boolean isPostingEntryRepo() {
		return getType() != null && getType() == GrilleType.POSTING_ENTRY_REPOSITORY;
	}
	
	@JsonIgnore
	public boolean isBookingRepo() {
		return getType() != null && getType() == GrilleType.BOOKING_REPOSITORY;
	}
	
	@JsonIgnore
	public boolean isArchiveBackup() {
		return getType() != null && getType() == GrilleType.ARCHIVE_BACKUP;
	}
	
	@JsonIgnore
	public boolean isArchiveReplacement() {
		return getType() != null && getType() == GrilleType.ARCHIVE_REPLACEMENT;
	}
	
	@JsonIgnore
	public boolean isDashboardReport() {
		return getType() != null && getType() == GrilleType.DASHBOARD_REPORT;
	}
	
	@JsonIgnore
	public boolean isReporting() {
		return isDashboardReport() || isReport() || isReconciliation() || isBillingEventRepo() || isClientRepo() || isPostingEntryRepo() || isBookingRepo() || isArchiveBackup() || isArchiveReplacement();
	}
	
	

	

	@PostLoad
	public void initListChangeHandler() {
		columns.forEach( item -> { });	
		dimensions.forEach( item -> { });		
		this.columnListChangeHandler.setOriginalList(columns);
		this.dimensionListChangeHandler.setOriginalList(dimensions);
	}
	
	public GrilleColumn getColumnAt(int position) {
		for(GrilleColumn column : getColumnListChangeHandler().getItems()) {
			if(position == column.getPosition()) {
				return column;
			}
		}
		return null;
	}
	
	public GrilleColumn getColumnById(Long id) {
		for (GrilleColumn column : this.getColumnListChangeHandler().getItems()) {
			if (column.getId().equals(id)) {				
				return column;
			}
		}
		return null;
	}
	
	public List<GrilleColumn> getColumns(DimensionType dimensionType, Long dimensionId) {
		List<GrilleColumn> columns = new ArrayList<>();
		for(GrilleColumn column : getColumnListChangeHandler().getItems()) {
			if(dimensionType == column.getType() && dimensionId.equals(column.getDimensionId())) {
				columns.add(column);
			}
		}
		return columns;
	}
	
	public GrilleColumn getColumnByDimensionAndName(DimensionType dimensionType, String name) {
		for(GrilleColumn column : getColumnListChangeHandler().getItems()) {
			if(column.getType() == dimensionType && column.getName() != null && column.getName().equals(name)) {
				return column;
			}
		}
		return null;
	}
	
	@JsonIgnore
	public String getPublicationTableName() {
		return "GRID_" + getId();
	}
		

	@JsonIgnore
	@Override
	public Grille copy() {
		Grille copy = new Grille();
		copy.setName(this.getName() + System.currentTimeMillis());
		copy.setDataSourceId(getDataSourceId());
		copy.setDataSourceType(getDataSourceType());
		copy.setDataSourceName(getDataSourceName());
		copy.setGroup(this.getGroup());
		copy.setVisibleInShortcut(isVisibleInShortcut());
		copy.setShowAllRowsByDefault(isShowAllRowsByDefault());
		copy.setStatus(getStatus());
		copy.setType(getType());
		copy.setSourceType(getSourceType());
		copy.setSourceId(getSourceId());
		copy.setReportId(getReportId());
		copy.setCategory(getCategory());
		copy.setEditable(isEditable());	
		copy.setAllowLineCounting(isAllowLineCounting());
		copy.setConsolidated(isConsolidated());
		copy.setDebit(isDebit());
		copy.setCredit(isCredit());
		copy.setRowType(getRowType());
		copy.setPublished(getPublished());
		copy.setUserFilter(userFilter != null ? userFilter.copy() : null);
		copy.setAdminFilter(adminFilter != null ? adminFilter.copy() : null);
		for(GrilleColumn column : getColumnListChangeHandler().getItems()) {
			copy.getColumnListChangeHandler().addNew(column.copy());
		}
		for(GrilleDimension dimension : getDimensionListChangeHandler().getItems()) {
			copy.getDimensionListChangeHandler().addNew(dimension.copy());
		}		
		return copy;
	}



}

File: client/bcephal-client-model/Grids/Grille.cs
using Bcephal.Models.Base;
using Bcephal.Models.Filters;
using Bcephal.Models.Links;
using Bcephal.Models.Utils;
using Newtonsoft.Json;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading.Tasks;

namespace Bcephal.Models.Grids
{
	[Serializable]
	public class Grille : MainObject
	{
        public GrilleType Type { get; set; }

        public GrilleCategory Category { get; set; }

        public GrilleStatus Status { get; set; }

        public GrilleSource SourceType { get; set; }

        public long? SourceId { get; set; }

        public bool Editable { get; set; }

        public bool Consolidated { get; set; }

        public bool ShowAllRowsByDefault { get; set; }

        public bool AllowLineCounting { get; set; }

        public bool UseLink { get; set; }

        public bool? Debit { get; set; }

        public bool? Credit { get; set; }
        public bool Published { get; set; }

        public string RowType { get; set; }
        [JsonIgnore]
        public GrilleRowType GrilleRowType
        {
            get
            {
                return string.IsNullOrEmpty(RowType) ? GrilleRowType.ALL : GrilleRowType.GetByCode(RowType);
            }
            set
            {
                this.RowType = value != null ? value.getCode() : null;
            }
        }

        public UniverseFilter UserFilter { get; set; }

        public UniverseFilter AdminFilter { get; set; }

        public ListChangeHandler<GrilleDimension> DimensionListChangeHandler { get; set; }

        public ListChangeHandler<GrilleColumn> ColumnListChangeHandler { get; set; }

        public ListChangeHandler<Link> LinkListChangeHandler { get; set; }


        [JsonIgnore]
        public bool IsLoaded
        {
            get
            {
                return this.Status == GrilleStatus.LOADED;
            }
            set
            {
                this.Status = value ? GrilleStatus.LOADED : GrilleStatus.UNLOADED;
            }
        }

        [JsonIgnore]
        public bool IsUnLoaded
        {
            get
            {
                return this.Status == GrilleStatus.UNLOADED;
            }
            set
            {
                this.Status = value ? GrilleStatus.UNLOADED : GrilleStatus.LOADED;
            }
        }

        [JsonIgnore] public bool IsReconciliation
        {
            get { return this.Type == GrilleType.RECONCILIATION; }
        }

        [JsonIgnore] public bool IsReport
        {
            get { return this.Type == GrilleType.REPORT; }
        }

        [JsonIgnore]  public bool IsInput
        {
            get { return this.Type == GrilleType.INPUT; }
        }

        [JsonIgnore]
        public bool IsBillingEventRepo
        {
            get { return this.Type == GrilleType.BILLING_EVENT_REPOSITORY; }
        }

        [JsonIgnore]
        public bool IsClientRepo
        {
            get { return this.Type == GrilleType.CLIENT_REPOSITORY; }
        }

        [JsonIgnore]
        public bool IsBookingRepo
        {
            get { return this.Type == GrilleType.BOOKING_REPOSITORY; }
        }

        [JsonIgnore]
        public bool IsPostingEntryRepo
        {
            get { return this.Type == GrilleType.POSTING_ENTRY_REPOSITORY; }
        }

        [JsonIgnore]
        public bool IsForReporting
        {
            get { return this.IsReport || this.IsReconciliation || this.IsBillingEventRepo || this.IsBookingRepo || this.IsPostingEntryRepo; }
        }

        public Grille()
        {
            this.ColumnListChangeHandler = new ListChangeHandler<GrilleColumn>();
            this.LinkListChangeHandler = new ListChangeHandler<Link>();
            this.DimensionListChangeHandler = new ListChangeHandler<GrilleDimension>();
        }

        public Grille(Grille grid, bool copyFilter = false) : this()
        {
            this.Id = grid.Id;
            this.Name = grid.Name;
            this.ColumnListChangeHandler.SetOriginalList(grid.ColumnListChangeHandler.Items);
            this.UseLink = grid.UseLink;
            this.Consolidated = grid.Consolidated;
            this.Type = grid.Type;
            this.Status = grid.Status;
            this.Debit = grid.Debit;
            this.Credit = grid.Credit;
            this.GrilleRowType = grid.GrilleRowType;
            if (copyFilter)
            {
                this.UserFilter = grid.UserFilter;
                this.AdminFilter = grid.AdminFilter;
            }
        }


        public void AddColumn(GrilleColumn column, bool sort = true)
        {
            column.Position = ColumnListChangeHandler.Items.Count;
            ColumnListChangeHandler.AddNew(column, sort);
        }

        public void UpdateColumn(GrilleColumn column, bool sort = true)
        {
            ColumnListChangeHandler.AddUpdated(column, sort);
        }

        public void InsertColumn(int position, GrilleColumn column)
        {
            column.Position = position;
            foreach (GrilleColumn child in ColumnListChangeHandler.Items)
            {
                if (child.Position >= column.Position)
                {
                    child.Position = child.Position + 1;
                    ColumnListChangeHandler.AddUpdated(child, false);
                }
            }
            ColumnListChangeHandler.AddNew(column);
        }


        public void DeleteOrForgetColumn(GrilleColumn column)
        {
            if (column.IsPersistent)
            {
                DeleteColumn(column);
            }
            else
            {
                ForgetColumn(column);
            }
        }

        public void DeleteColumn(GrilleColumn column)
        {
            ColumnListChangeHandler.AddDeleted(column);
            foreach (GrilleColumn child in ColumnListChangeHandler.Items)
            {
                if (child.Position > column.Position)
                {
                    child.Position = child.Position - 1;
                    ColumnListChangeHandler.AddUpdated(child, false);
                }
            }
        }

        public void ForgetColumn(GrilleColumn column)
        {
            ColumnListChangeHandler.forget(column);
            foreach (GrilleColumn child in ColumnListChangeHandler.Items)
            {
                if (child.Position > column.Position)
                {
                    child.Position = child.Position - 1;
                    ColumnListChangeHandler.AddUpdated(child, false);
                }
            }
        }


        public GrilleColumn GetColumnByDimensionId(long? dimensionId, DimensionType type)
        {
            foreach (GrilleColumn column in ColumnListChangeHandler.Items)
            {
                if (column.DimensionId.Equals(dimensionId) && type == column.Type)
                {
                    return column;
                }
            }
            return null;
        }

        public List<GrilleColumn> GetPersistentColumns(string type = null)
        {
            List<GrilleColumn> columns = new List<GrilleColumn>();
            foreach (GrilleColumn column in ColumnListChangeHandler.Items)
            {
                if (column.IsPersistent)
                {
                    if (type != null)
                    {
                        if (type.Equals(column.Type.ToString()))
                        {
                            columns.Add(column);
                        }
                    }
                    else
                    {
                        columns.Add(column);
                    }
                }
            }
            return columns;
        }




        public void AddLink(Link link, bool sort = true)
        {
            link.Position = ColumnListChangeHandler.Items.Count;
            LinkListChangeHandler.AddNew(link, sort);
        }
        public void UpdateLink(Link link, bool sort = true)
        {
            LinkListChangeHandler.AddUpdated(link, sort);
        }

        public void DeleteOrForgetLink(Link link)
        {
            if (link.IsPersistent)
            {
                DeleteLink(link);
            }
            else
            {
                ForgetLink(link);
            }
        }

        public void DeleteLink(Link link)
        {
            LinkListChangeHandler.AddDeleted(link);
            foreach (Link child in LinkListChangeHandler.Items)
            {
                if (child.Position > link.Position)
                {
                    child.Position = child.Position - 1;
                    LinkListChangeHandler.AddUpdated(child, false);
                }
            }
        }

        public void ForgetLink(Link link)
        {
            LinkListChangeHandler.forget(link);
            foreach (Link child in LinkListChangeHandler.Items)
            {
                if (child.Position > link.Position)
                {
                    child.Position = child.Position - 1;
                    LinkListChangeHandler.AddUpdated(child, false);
                }
            }
        }


        public void AddDimension(GrilleDimension dimension, bool sort = true)
        {
            dimension.Position = DimensionListChangeHandler.Items.Count;
            DimensionListChangeHandler.AddNew(dimension, sort);
        }

        public void UpdateDimension(GrilleDimension dimension, bool sort = true)
        {
            DimensionListChangeHandler.AddUpdated(dimension, sort);
        }


        public void DeleteOrForgetDimension(GrilleDimension dimension)
        {
            if (dimension.IsPersistent)
            {
                DeleteDimension(dimension);
            }
            else
            {
                ForgetDimension(dimension);
            }
        }

        public void DeleteDimension(GrilleDimension dimension)
        {
            DimensionListChangeHandler.AddDeleted(dimension);
            foreach (GrilleDimension child in DimensionListChangeHandler.Items)
            {
                if (child.Position > dimension.Position)
                {
                    child.Position = child.Position - 1;
                    DimensionListChangeHandler.AddUpdated(child, false);
                }
            }
        }

        public void ForgetDimension(GrilleDimension dimension)
        {
            DimensionListChangeHandler.forget(dimension);
            foreach (GrilleDimension child in DimensionListChangeHandler.Items)
            {
                if (child.Position > dimension.Position)
                {
                    child.Position = child.Position - 1;
                    DimensionListChangeHandler.AddUpdated(child, false);
                }
            }
        }


        public GrilleDimension GetDimensionByDimensionId(long? dimensionId, DimensionType type)
        {
            foreach (GrilleDimension dimension in DimensionListChangeHandler.Items)
            {
                if (dimension.DimensionId.Equals(dimensionId) && type == dimension.Type)
                {
                    return dimension;
                }
            }
            return null;
        }









        public GrilleColumn GetColumn(string fieldName)
        {
            foreach (GrilleColumn column in this.ColumnListChangeHandler.Items)
            {
                if (column.Name.Equals(fieldName))
                {
                    return column;
                }
            }
            return null;
        }

        public GrilleColumn GetColumn(DimensionType type, long? id)
        {
            foreach (GrilleColumn column in this.ColumnListChangeHandler.Items)
            {
                if (column.Type == type && column.DimensionId == id) return column;
            }
            return null;
        }


        public void  reverseColumn(int previewPosition, int NewPosition, GrilleColumn column)
        {
            bool isUp =  previewPosition > NewPosition;
            ObservableCollection<GrilleColumn> its_ = ColumnListChangeHandler.Items;
            its_.BubbleSort();
            IList<GrilleColumn> its = its_;
            if (NewPosition >= 0 && NewPosition < its.Count && previewPosition >= 0 && previewPosition < its.Count)
            {
                if (isUp)
                {
                    GrilleColumn child = its[NewPosition];
                    column.Position = NewPosition;
                    ColumnListChangeHandler.AddUpdated(column, false);
                    its[NewPosition] = column;
                    int position_ = previewPosition;
                    while (position_ - 1 > NewPosition)
                    {
                        its[position_] = its[position_ - 1];
                        its[position_].Position = position_;
                        ColumnListChangeHandler.AddUpdated(its[position_], false);
                        position_--;
                    }
                    its[position_] = child;
                    child.Position = position_;
                    ColumnListChangeHandler.AddUpdated(child, false);
                }
                else
                {
                    GrilleColumn child = its[NewPosition];
                    column.Position = NewPosition;
                    ColumnListChangeHandler.AddUpdated(column, false);
                    its[NewPosition] = column;
                    int position_ = previewPosition;
                    while (position_ + 1 < NewPosition)
                    {
                        its[position_] = its[position_ + 1];
                        its[position_].Position = position_;
                        ColumnListChangeHandler.AddUpdated(its[position_], false);
                        position_++;
                    }
                    its[position_] = child;
                    child.Position = position_;
                    ColumnListChangeHandler.AddUpdated(child, false);
                }
            }
        }
    }
}
